### ISSUE TITLE
Implement attachement of code comments to AST node via an AST attribute system

### ISSUE THREAD

**Issue by denismerigoux** [CatalaLang/catala#736]
In order to make progress on #673, the attribute system of #392 has entered the critical path. This issue states that work on the attribute system should begin with the following objectives : 
* capture comments inside Catala code blocks as either surfact AST nodes or annotations on neighbouring surface AST nodes (the parser and lexer should be modified) ;
* the rules for knowing where to attach comments to AST nodes can be simple at first : the comment is always attached to the AST node of the token immediately after the comment in the linear flow of tokens ;
* code comments attached to AST nodes should be pushed down the compilation chain at least to Desugared (used by #673).

**Comment by AltGr** [comment]
Proposed attribute syntax (similar to [Rust](https://doc.rust-lang.org/reference/attributes.html), itself taken from C#):
- outer syntax `#[ATTR]`
- always bound to the AST node immediately following
- (optionally, we could also support `#![ATTR]` that bind to their parent node)
- inner syntax: `ATTR` is the form `VAR`, or `VAR = VAL`, where
  - `VAR` is a lident, or a dot-separated list thereof (`path.to.key`)
  - `VAL` is a `"`-delimited string (where `\"` escapes `"` and `\\` escapes `\` ?)
  - (optionally, we could extend that to also support arbitrary literals, or even arbitrary expressions ?)

example:
```catala-fr
declaration structure InputStruct:
  #[form.user_description = "This field corresponds to XXX and should be filled if YYY"]
  data user_xxx_data content money

#[test]
declaration scope Test1:
  ...
```

Implementation-wise, they will be added to the marks which already contain position and optionally type data

**Comment by AltGr** [comment]
Some implementation thoughts: at the moment, extended marks are only present on expressions. Annotations will be useful on types, constructors, declarations, etc. as well.

Since these already have positions, it is tempting to simply extend the positions with annotation data ; or replace them with a hierarchical type, but this will require an update to all the places were positoin data is accessed.
Also: annotations themselves will have (raw) positions

**Comment by denismerigoux** [comment]
I like your specification ! I'm also in favor of just extending the position marks everywhere to contain attributes. And yes the attribute type should carry the positions of where the attributes are in the source code :)

**Comment by denismerigoux** [comment]
Good catch, Pierre. Yes, definitely, and with attributes the system could be more flexible and customizable than with hardcoded syntax additions.

**Comment by AltGr** [comment]
So that would be something like:
```catala-code-en
scope S:
  #[requires]
  assert input1 >= 0

  #[ensures]
  assert output1 <= input1
```

(or possibly ?
```catala-code-en
scope S:
  assert #![requires] input1 >= 0
```
with the inner-attribute syntax)

The positive side I see with this approach is that they all remain standard assertions, but the tagged ones may be picked up for special purposes :+1:
